lib/compiler/container.ts

```diff
@@ -0,0 +1,161 @@
+import { ClassDeclaration, SyntaxKind, TypeGuards } from 'ts-morph';
+import { Compiler } from '.';
+import { container } from '../tsdi';
+import { Component } from './component';
+import { UnitImpl } from './unit';
+import { findDeclarationForIdentifier } from './util';
+
+export class Container<T> {
+  public get name(): string {
+    return this.clazz.getNameOrThrow();
+  }
+
+  public get implName(): string {
+    return `${this.name}Impl`;
+  }
+
+  public get entryPoints(): { name: string; type: Component }[] {
+    const properties = this.clazz
+      .getProperties()
+      .filter(property => property.isAbstract());
+
+    return properties.map(property => {
+      const identifier = property
+        .getTypeNodeOrThrow()
+        .getFirstChildByKindOrThrow(SyntaxKind.Identifier);
+
+      const node = findDeclarationForIdentifier(identifier);
+      if (
+        TypeGuards.isClassDeclaration(node) ||
+        TypeGuards.isInterfaceDeclaration(node)
+      ) {
+        return {
+          name: property.getName(),
+          type: new Component(this, node)
+        };
+      }
+      throw new Error('Illegal node type for component: ' + node.print());
+    });
+  }
+
+  private _units?: UnitImpl[];
+
+  // tslint:disable-next-line:cyclomatic-complexity
+  public get units(): UnitImpl[] {
+    if (this._units) {
+      return this._units;
+    }
+    const containerDecorator = this.clazz.getDecorator(container.name);
+    if (containerDecorator && containerDecorator.isDecoratorFactory()) {
+      const options = containerDecorator.getArguments()[0];
+      if (!TypeGuards.isObjectLiteralExpression(options)) {
+        throw new Error('Invalid container decorator: ' + this.clazz.print());
+      }
+      const unitsProperty = options.getProperty('units');
+      if (!unitsProperty || !TypeGuards.isPropertyAssignment(unitsProperty)) {
+        throw new Error('Invalid container units: ' + this.clazz.print());
+      }
+      const unitsArray = unitsProperty.getInitializerIfKindOrThrow(
+        SyntaxKind.ArrayLiteralExpression
+      );
+      this._units = unitsArray.getElements().map(element => {
+        if (!TypeGuards.isIdentifier(element)) {
+          throw new Error(
+            'Invalid container unit must be identifier: ' + element.print()
+          );
+        }
+        const node = findDeclarationForIdentifier(element);
+        if (TypeGuards.isClassDeclaration(node)) {
+          return new UnitImpl(this, node);
+        } else {
+          throw new Error('Illegal node type for unit: ' + node.print());
+        }
+      });
+    } else {
+      this._units = [];
+    }
+    return this._units;
+  }
+
+  private get internals(): Component[] {
```

Comment:
Right. `internals` is every managed piece which is not `external` :smile: 
Internals and externals are a fragment of transitioning from version 0.x where externals where components which need to have injections but could not be managed by TSDI.
This concept is now hidden for most of the users of TSDI.

There are cases where this will be relevant.
One of them would be having multiple container as parts of your dependency tree and having `@managed` classes which are not directly integrated in a managed tree of one of the containers. Then users need to flag which container should managed this piece. And to create a meaningful error message this internal-external thing is there.